cpu_rec.py

```diff
@@ -77,37 +31,33 @@
     log.addHandler(console_handler)
     log.setLevel(logging.WARN)
 
-# NB: we get a string in python2 and bytes in python3
-if sys.version_info[0] == 2:
-    byte_ord = lambda i: ord(i)
-else:
-    byte_ord = lambda i: i
-
 class TrainingData(object):
     def __init__(self):
-        """ This class contains, for each item in the training corpus:
-            architecture, file name, binary data """
+        """This class contains, for each item in the training corpus:
+        architecture, file name, binary data"""
         self.archs = []
         self.files = []
         self.data = []
+
     def dump(self, dumpdir=None):
-        """ Dump the raw corpus, in a form that won't need elftesteem to be loaded """
+        """Dump the raw corpus, in a form that won't need elftesteem to be loaded"""
         for arch, data in zip(self.archs, self.data):
-            of = open(dumpdir+'/'+arch.replace('/','-')+'.corpus', 'ab')
+            of = open(dumpdir + '/' + arch.replace('/', '-') + '.corpus', 'ab')
             of.write(data)
             of.close()
+
     def add_training(self, arch, file=None, section='text', data=None, repeat=1):
-        """ Add a new item in the training corpus:
-            'arch': architecture name
-            if 'data' is not None:
-                'data': raw bytestream
-            else:
-                'file': file to read
-                'section': part of the file to extract
-                    None: keep the whole file
-                    of type 'slice': start and end of part to extract
-                    of type 'str': name of the section to extract
-            'repeat': when the corpus is too small, we repeat it
+        """Add a new item in the training corpus:
+        'arch': architecture name
+        if 'data' is not None:
+            'data': raw bytestream
+        else:
+            'file': file to read
+            'section': part of the file to extract
+                None: keep the whole file
+                of type 'slice': start and end of part to extract
+                of type 'str': name of the section to extract
+        'repeat': when the corpus is too small, we repeat it
         """
         if data is None:
             data = TrainingData.unpack_file(open(file, 'rb').read())
```

Comment:
Leak of the resource and waste of memory. Use context manager to prevent the leak. And, maybe, a `mmap`, to use memory more efficiently.

Reply:
The file descriptor is not leaked, as the reference is deleted right after the read.